### ISSUE TITLE
Improve the performance of sqlfluff

### ISSUE THREAD

**Issue by dclong** [sqlfluff/sqlfluff#962]
I know that there has been effort to improve the performance of sqlfluff. However, sqlfluff still runs extremely slow. For example, I have a SQL file (containing about 3000 lines of code after formatted using sqlfluff) and it takes about 1 hour to lint it and close to 3 hours to run `sqlfluff fix` on it.

**Comment by barrywhart** [comment]
This is more of a question or suggestion than a solution, but have we tried running a big SQLFluff job using PyPy rather than CPython? For CPU-bound applications, PyPy is often 5-7x faster. I've used it on a handful of projects over the past several years. Sometimes there are compatibility issues, but that's gotten much better recently.

https://www.pypy.org/

**Comment by dmateusp** [comment]
@dclong would it be possible for you to remove any sensitive information from that 3000 line SQL file and submit it to that issue? (I know it's a big ask, it's a lot of SQL!)

Then we could do some profiling using your query directly

**Comment by alanmcruickshank** [comment]
Thanks @dclong . As options for what we do here:
- As suggested by @barrywhart , I've just checked and sqlfluff does run out of the box on pypy (using `pypy -m sqlfluff ...`). I haven't gone deep into functionality but everything looks good so far. On your large project I would definitely be interested on the performance gains you see there.
- In the last release (`0.5.2`) there were some logging improvements to see explicitly where the timing was going, especially for the dbt templater. Can you run the command you normally run but with `-vv` on the end to see the how long the compilation step takes?
- If neither of these are the issue then I have a feeling that this is going to be the `rule` evaluation process. I've noticed this on our codebase too which has a log of sql, and while it doesn't take as long as yours, it certainly takes a long time. I think it might be worth us re-evaluating the way that rules works to enable them to work faster. In particular at the moment, the application of rules is really expensive at the moment in that they're called on *every* segment in the parsed structure, and the rule then skips the segment if it's not relevant. There are definitely options to improve this.
- To potentially confirm this, when you run `fix` can you add `--bench` to the command which should output a section at the end with a readout of where sqlfluff was spending it's time. Unfortunately most of the benchmarking code is focussed on the _parsing_ elements of the process and so we might not get much granularity on the rules element if that is the culprit.
- For the very large SQL file, if you're not comfortable posting it on the issue here, I'm happy for you to DM me on the sqlfluff slack or email it to me for me to have a look at and unpick where the performance issues are.

**Comment by dclong** [comment]
I tried to run sqlfluff using PyPy. Linting was about 2 times faster however `sqlfluff fix` was even slower. I will try using `--bench` to see which part takes the most time. 

Unfortunately, I couldn't post the SQL code here but I will try to anonymize it and post the anonymized version later.

**Comment by barrywhart** [comment]
I found a way to use native Python profiling, rather than the SQLFluff profliing. This provides a more detailed, function-call-level view of the running time without having to share confidential SQL files.

@dclong: In the project directory where you normally run the `sqlfluff` command, create a new file `sqlfluff_cli.py` with the following contents:
```
import sys

from sqlfluff.cli.commands import cli

cli(sys.argv[1:])

```

Then run this command (replacing `sql/user_stats/user_stats_combined.sql` with your 3000-line SQL file):
```
python -m cProfile -s cumtime sqlfluff_cli.py lint sql/user_stats/user_stats_combined.sql > output.txt
```

This will produce a file `output.txt` that contains the SQLFluff output followed by a detailed report telling where SQLFluff is spending its time. Please upload this file to the issue. I ran this on one of our larger files, and I will upload mine here as well.

I did a preliminary analysis of my profile report, and **two rules really dominate the running time: L010 and L014**. These are both capitalization-related rules, and L014 is a subclass of L010, so for practical purposes, they are basically the same rule.

@dclong: **Can you also try linting this file (and your whole project) with L010 and L014 disabled?** This may give you a nice performance boost for now, while we are working on improving performance.

**Comment by barrywhart** [comment]
This is the `output.txt` file from linting/profiling `sql/user_stats/user_stats_combined.sql`, a ~150-line SQL file in one of our projects.

[output.txt](https://github.com/sqlfluff/sqlfluff/files/6331241/output.txt)

**Comment by barrywhart** [comment]
I will take a look at speeding up rules 10 and 14. I won't assign the ticket to myself for now, at least not until we confirm this is the cause of @dclong's slow performance as well.

**Comment by dclong** [comment]
@barrywhart Please see attached. 
[profile.txt](https://github.com/sqlfluff/sqlfluff/files/6332016/profile.txt)

**Comment by barrywhart** [comment]
Thanks, @dclong! I briefly reviewed the profile you uploaded. There are some similarities to mine, in that L010 and L014 are the slowest rules in each. However, for your file, it appears the parser is the slowest part, rather than the rules.

I created a PR, #983, which significantly speeds up L010 and L014. Feel free to test that if you like. I think you will see a meaningful but not huge improvement in performance.

I will also work with @alanmcruickshank to see if he or I wants to look into parser performance as well.

**Comment by barrywhart** [comment]
From an initial look at the parser, tuning it could be a more involved, long-term project. I say this partly because the parser spends its time in a number of `match()` methods which call one another recursively, based on the structure of the grammar and the SQL code. This makes "top down" profiling (look for the method where the most cumulative time is spent) not very productive. Instead, we'd likely want to look at "bottom up" tuning, i.e. find the methods with the highest "self time", i.e. time spent in that method, not the children, then see if we can either:
* Speed up those methods
* Think of clever ways to reduce calls to that method (redesign, caching, etc.)

For a quicker hit, we may want to look at #642, i.e. parallelizing `sqlfluff lint` and `sqlfluff fix`. This could result in a fairly linear speedup based on the number of CPUs.

**Comment by dclong** [comment]
@dmateusp @alanmcruickshank Attached is the anonymized version of the SQL. 
[anynomize.sql.txt](https://github.com/sqlfluff/sqlfluff/files/6332578/anynomize.sql.txt)

**Comment by barrywhart** [comment]
#642 has been addressed, adding support for running lint and fix on multiple processors. This should help with performance issues. Leaving this issue open, though, as I think there are still some good opportunities to improve raw performance. Also see #1046 for another proposed performance improvement.

**Comment by dclong** [comment]
Does the `--parallel` option improve performance only when there are multiple SQL files? I tried the option on the attached SQL file `anonymize.sql.txt` but it didn't speed up linting.

**Comment by barrywhart** [comment]
Correct -- individual files are processed the same as before, but if there are many files, it can process more than one at a time. In theory, it could be X times faster, where X is the number you specify to the `--parallel` option. Currently, it works okay on Linux and MacOS. We're working on some issues with Windows. Performance will hopefully improve as we do more testing and tweaks.

**Comment by dclong** [comment]
Is it possible to parse a single SQL file as an execution graph and then leverage parallel computing? This way we can improve the performance on single SQL file too.

**Comment by barrywhart** [comment]
I'm going to close this. We:
* Addressed some rule performance issues in #983 
* Added parallel linting in #642
* Have a new issue (#1208) focused on slow parsing performance for a _single_ file. (This has not been addressed yet.)
